cmd/dp-observation-importer/main.go

```diff
@@ -82,69 +159,137 @@ func main() {
+	log.Event(ctx, fmt.Sprintf("Shutdown with timeout: %s", cfg.GracefulShutdownTimeout), log.INFO)
+	shutdownContext, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
 
-	// gracefully dispose resources
-	err = eventConsumer.Close(ctx)
-	logError(err)
+	// Gracefully shutdown the application closing any open resources.
+	go func() {
+		defer cancel()
 
-	err = kafkaConsumer.Close(ctx)
-	logError(err)
+		if serviceList.HealthCheck {
+			hc.Stop()
+		}
 
-	err = kafkaErrorProducer.Close(ctx)
-	logError(err)
+		err = httpServer.Shutdown(shutdownContext)
+		logIfError(ctx, "failed to shutdown http server", err)
 
-	err = kafkaResultProducer.Close(ctx)
-	logError(err)
+		// If kafka consumer exists, stop listening to it. (Will close later)
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "stopping kafka consumer listener", log.INFO)
+			syncConsumerGroup.StopListeningToConsumer(shutdownContext)
+			log.Event(shutdownContext, "stopped kafka consumer listener", log.INFO)
+		}
 
-	err = graphDB.Close(ctx)
-	logError(err)
+		// If observation imported kafka producer exists, close it
+		if serviceList.ObservationsImportedProducer {
+			log.Event(shutdownContext, "closing observation imported kafka producer", log.INFO, log.Data{"producer": "DimensionExtracted"})
+			observationsImportedProducer.Close(shutdownContext)
+			log.Event(shutdownContext, "closed observation imported kafka producer", log.INFO, log.Data{"producer": "DimensionExtracted"})
+		}
 
-	err = httpServer.Shutdown(ctx)
-	logError(err)
+		// If observation imported error kafka producer exists, close it
+		if serviceList.ObservationsImportedErrProducer {
+			log.Event(shutdownContext, "closing observation imported error kafka producer", log.INFO, log.Data{"producer": "DimensionExtracted"})
+			observationsImportedErrProducer.Close(shutdownContext)
+			log.Event(shutdownContext, "closed observation imported error kafka producer", log.INFO, log.Data{"producer": "DimensionExtracted"})
+		}
 
-	// cancel the timer in the shutdown context.
-	cancel()
+		// Attempting to close event consumer
+		log.Event(shutdownContext, "closing event wrapper", log.INFO)
+		eventConsumer.Close(shutdownContext)
+		log.Event(shutdownContext, "closing event wrapper", log.INFO)
 
-	log.Debug("graceful shutdown was successful", nil)
-	os.Exit(0)
+		// If kafka consumer exists, close it.
+		if serviceList.Consumer {
+			log.Event(shutdownContext, "closing kafka consumer", log.INFO, log.Data{"consumer": "SyncConsumerGroup"})
+			syncConsumerGroup.Close(shutdownContext)
+			log.Event(shutdownContext, "closed kafka consumer", log.INFO, log.Data{"consumer": "SyncConsumerGroup"})
+		}
+
+		if serviceList.Graph {
+			err = graphDB.Close(ctx)
+			logIfError(ctx, "failed to close graph db", err)
+		}
+	}()
+
+	// wait for shutdown success (via cancel) or failure (timeout)
+	<-shutdownContext.Done()
+
+	log.Event(shutdownContext, "graceful shutdown was successful", log.INFO)
```

Comment:
If a failure occurs on graceful shutdown, the log message will imply that it was still a success?